File: backend/UMS/UMS.Domain/Entities/Grade.cs
using System;
using UMS.Domain.Enums;

namespace UMS.Domain.Entities
{
    public class Grade
    {
        public Guid Id { get; private set; }
        public Guid EnrollmentId { get; private set; }
        public GradeType GradeType { get; private set; }
        public float Score { get; private set; }
        public float MaxScore { get; private set; }
        public float Weight { get; private set; }
        public Guid? UpdatedBy { get; private set; }
        public DateTime GradedAt { get; private set; }
        public string? Note { get; private set; }

        public Enrollment? Enrollment { get; private set; }

        private Grade() { }

        public Grade(Guid enrollmentId, GradeType gradeType, float score, Guid updatedBy, float maxScore = 10, float weight = 1.0f, string? note = null)
        {
            Id = Guid.NewGuid();
            EnrollmentId = enrollmentId;
            GradeType = gradeType;
            Score = score;
            MaxScore = maxScore;
            Weight = weight;
            UpdatedBy = updatedBy;
            Note = note;
            GradedAt = DateTime.UtcNow;
        }

        public void UpdateScore(float score, Guid updatedBy, string? note = null)
        {
            Score = score;
            UpdatedBy = updatedBy;
            Note = note;
            GradedAt = DateTime.UtcNow;
        }
    }
}

File: frontend/ums/src/types/grade.ts
// types/grade.ts

export enum GradeType {
  Attendance = 0,
  Midterm = 1,
  Final = 2,
  Assignment = 3,
}

export const GradeTypeLabel: Record<GradeType, string> = {
  [GradeType.Attendance]: 'Chuyên cần',
  [GradeType.Midterm]: 'Giữa kỳ',
  [GradeType.Final]: 'Cuối kỳ',
  [GradeType.Assignment]: 'Bài tập',
}

// Loại điểm được phép nhập tay (Attendance chỉ tự sinh qua Sync, không nhập tay)
export const manualGradeTypeOptions = [
  { label: GradeTypeLabel[GradeType.Assignment], value: GradeType.Assignment },
  { label: GradeTypeLabel[GradeType.Midterm], value: GradeType.Midterm },
  { label: GradeTypeLabel[GradeType.Final], value: GradeType.Final },
]

// ─── Response DTO ───
export interface GradeResponse {
  id: string
  enrollmentId: string
  studentId: string
  studentFullName: string
  gradeType: GradeType
  score: number
  maxScore: number
  weight: number
  gradedAt: string
  note?: string | null
}

export interface FinalGradeResponse {
  enrollmentId: string
  studentId: string
  studentFullName: string
  finalScore: number
  gradeLetter: string
}

// ─── Request DTO ───
export interface CreateGradeRequest {
  enrollmentId: string
  gradeType: GradeType
  score: number
  maxScore?: number
  weight?: number
  note?: string
}

export interface UpdateGradeRequest {
  score: number
  note?: string
}
